pubtools/_content_gateway/push_staged.py

```diff
@@ -1,2 +1,191 @@
+from pushsource import Source, CGWPushItem
+from .cgw_client import *
+from .cgw_authentication import CGWBasicAuth
+from .utils import yaml_parser, validate_schema
+import logging
+
+LOG = logging.getLogger("pubtools.cgw")
+LOG_FORMAT = "%(asctime)s [%(levelname)-8s] %(message)s"
+logging.basicConfig(level=logging.DEBUG, format=LOG_FORMAT)
+
+
+class PushStagedCGW:
+    def __init__(self, push_items, hub, task_id, target_name, target_settings):
+        self.auth = CGWBasicAuth(target_settings['username'], target_settings['password'])
+        self.cgw_client = CGWClient(target_settings['server_name'], self.auth)
+        self.product_mapping = None
+        self.pv_mapping = None
+
+        self.push_items = push_items
+        self.hub = hub
+        self.task_id = task_id
+        self.target_name = target_name
+
+    def __create_product_version_mapping(self, product_name, product_code):
+        product_id = self.__get_product_id(product_name, product_code)
+        all_versions = self.cgw_client.get_versions(product_id)
+        self.pv_mapping = {(product_name, product_code, data.get('versionName')): (product_id, data.get('id')) for data
+                           in all_versions}
+
+    def __get_product_id(self, product_name, product_code):
+        product_id = self.product_mapping.get((product_name, product_code))
+        return product_id
+
+    def __get_version_id(self, product_name, product_code, version_name):
+        version_id = self.pv_mapping.get((product_name, product_code, version_name)) if self.pv_mapping else None
+        if version_id:
+            _, version_id = version_id
+        return version_id
+
+    def __get_file_id(self, file_item):
+        product_name = file_item.get('metadata').get('productName')
+        product_code = file_item.get('metadata').get('productCode')
+        version_name = file_item.get('metadata').get('productVersionName')
+
+        product_id = self.__get_product_id(product_name, product_code)
+        if not self.pv_mapping:
+            self.__create_product_version_mapping(product_name, product_code)
+        version_id = self.__get_version_id(product_name, product_code, version_name)
+        all_files = self.cgw_client.get_files(product_id, version_id)
+        file_mapping = {(product_name, product_code, version_name, data.get('downloadURL')): (
```

Comment:
Why doesn't this also have a create method similar to the other ones?

Reply:
Just because we will have to create file mapping every time when we'll do file operations and for products and versions, versions can use the product mapping same as File can use version mapping to save some time by eliminating extra API calls.
But for file, no further hierarchy is there and we have to keep looking the updated file mapping. So creating a separate function for file mapping will completely be useless.